### PR TITLE
[writeset-generator] Building the first release writeset

### PR DESCRIPTION
**Pull request by runtian-zhou**
## Motivation
This PR has two major components:
1. Update the rust `writeset-generator` with following changes:
    - Generate a writeset with a increment in DiemVersion. This would be needed for releasing the script function.
    -  Updated the release artifact file to make it append only. Also added other info needed for better reproducability.
2. Generate the actual writeset payload and artifact files for the `release-1.2.0` by using following command:
```
cargo run -- --output-payload -o release/release-1.2.0.blob create-release MAINNET https://mainnet.diem.com 13213663 --first-release --diem-version 2
```
Also changed the logic in versioning test to use this writeset blob to perform the versioning test in `e2e-testsuites`.

The second part of the PR is just for proving the writeset introduces no unexpected behavior. We can wait for the branch cut and land the second commit if needed.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes
## Test Plan

Golden files were changed for two reasons:
1. For `writeset_builder` tests, the writeop order was reversed, causing golden file to change.
2. For other golden files, it was because of the fact that we used one transaction to perform the module update and sequence number bump instead of two, causing diem_root's sequence number to be off by one.

### CODE REVIEW

**Comment by sblackshear** [comment]
> cargo run -- --output-payload -o release/release-1.2.0.blob create-release MAINNET https://mainnet.diem.com 13213663 --first-release --diem-version 2

Potentially dumb question, but why is `--first-release` required here? I thought this would be needed only for genesis. If we need to pass this flag every time, can we eliminate it?

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs [hunk 1]
```diff
@@ -35,29 +42,29 @@ pub fn create_release(
...
-        bail!("Artifact mismatch with on disk file");
-    }
-    if artifact.version > version {
-        bail!(
-            "Artifact version is ahead of the argument: old: {:?}, new: {:?}",
-            artifact.version,
-            version
-        );
+        if artifact.chain_id != chain_id {
+            bail!("Artifact mismatch with on disk file");
```
Comment (sblackshear) [inline language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:54]:
```suggestion
            bail!("Artifact chain id does not match input chain id. Artifact: {:?} Input {:?}", artifact.chain_id, chain_id);
```

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs [hunk 2]
```diff
@@ -35,29 +42,29 @@ pub fn create_release(
...
+        }
+        if artifact.version > version {
```
Comment (sblackshear) [inline language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:56]: Can we also check that `updated_diem_version > artifact.diem_version` (or perhaps better, `updated_diem_version == artifact.diem_version + 1`)?

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs [hunk 3]
```diff
@@ -75,7 +82,43 @@ pub(crate) fn create_release_from_artifact(
...
+    Ok(if let Some(updated_diem_version) = artifact.diem_version {
+        let state_view = DebuggerStateView::new(&remote, artifact.version);
+        let (updated_version_writeset, events) = build_changeset(&state_view, |session| {
+            session.set_diem_version(updated_diem_version);
+        })
+        .into_inner();
+
+        let (modules, _) = match modules_payload {
+            WriteSetPayload::Direct(cs) => cs,
+            _ => bail!("Unexpected AdminScript"),
```
Comment (sblackshear) [inline language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:96]:
```suggestion
            _ => bail!("Unexpected payload; wanted WriteSetPayload, found {:?}", payload),
```

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs [hunk 4]
```diff
@@ -75,7 +82,43 @@ pub(crate) fn create_release_from_artifact(
 ) -> Result<WriteSetPayload> {
     let remote = JsonRpcDebuggerInterface::new(remote_url)?;
     let remote_modules = remote.get_diem_framework_modules_by_version(artifact.version)?;
-    create_release_writeset(&remote_modules, release_modules)
+    let modules_payload = create_release_writeset(&remote_modules, release_modules)?;
+
+    Ok(if let Some(updated_diem_version) = artifact.diem_version {
```
[the remark this answers is not in the document]
Comment (sblackshear) [reply language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:87]: At one point, we discussed doing this (and similar tasks that require atomically running a script and updating modules) by allowing `AdminScript` to take a set of modules to publish. Is this idea still on the table?

**Comment by runtian-zhou** [comment]
> but why is --first-release required here?

The `--first-release` flag is just to tell the cli to bypass the check against previous release as this is our first time releasing a WriteSet patch. This is only required for this release and we can get rid of this flag in the future.

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs (hunk 2 above)
Comment (runtian-zhou) [reply language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:56]: We can't check for `updated_diem_version == artifact.diem_version + 1` as the AOS can increment this version on their end as well. Will add the test.

**Comment by runtian-zhou** [comment]
> allowing AdminScript to take a set of modules to publish. Is this idea still on the table?

This isn't implemented for now. Now with the ability of upgrading existing modules in module publishing function we might even update the frameworks with the regular modules. IMHO we should probably spend some time studying how we want the module publishing to work in the future.

**Comment by runtian-zhou** [comment]
Fixed @sblackshear's comments!

File: language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs (hunk 2 above)
Comment (sblackshear) [reply language/diem-tools/writeset-transaction-generator/src/release_flow/create.rs:60]: I see. Could we check against the current on-chain `diem_version` instead?

**Review by sblackshear — APPROVED** [review]

**Review by tzakian — APPROVED** [review]
LGTM, just one minor nit

**Pull request merged**